src/Control/Effect/Lift/Internal.hs

```diff
@@ -1,13 +1,21 @@
-{-# LANGUAGE DeriveFunctor, DeriveGeneric #-}
+{-# LANGUAGE ExistentialQuantification, RankNTypes #-}
 module Control.Effect.Lift.Internal
 ( Lift(..)
 ) where
 
 import Control.Effect.Class
-import GHC.Generics (Generic1)
+import Data.Functor.Compose
 
--- | @since 0.1.0.0
-newtype Lift sig m k = Lift { unLift :: sig (m k) }
-  deriving (Functor, Generic1)
+-- | @since 1.0.0.0
+data Lift sig m k
+  = forall a . LiftWith
+    (forall ctx . Functor ctx => ctx () -> (forall a . ctx (m a) -> sig (ctx a)) -> sig (ctx a))
+    (a -> m k)
 
-instance Functor m => Effect (Lift m)
+instance Functor m => Functor (Lift sig m) where
+  fmap f (LiftWith with k) = LiftWith with (fmap f . k)
+
+instance Functor sig => Effect (Lift sig) where
+  handle ctx dst (LiftWith with k) = LiftWith
+    (\ ctx' dst' -> getCompose <$> with (Compose (ctx <$ ctx')) (fmap Compose . dst' . fmap dst . getCompose))
+    (dst . fmap k)
```

Comment:
I initially based this branch off of #296, because I was initially only interested in providing a `MonadUnliftIO`-esque `((forall a . m a -> n a) -> n b) -> m b` operation, and that required us to have the constraint `ctx ~ Identity`. (Plus, it just seemed more tractable to begin with.)

However, after doing so, I wanted to see what it would take to do something more like `MonadBaseControl` (tho rephrased more like `Effect`’s own interface), where we capture and preserve the inner monad’s context; and that’s precisely the behaviour that `Effect` provides, just in different logical parts of the program: `Algebra` instances call `handle`, user code calls `liftWith`, but both do the same threading of handlers and contexts to unlift run code in an outer context and lift it back up into the inner one.

In a sense, this means that `Lift` is really the only effect you need, since you can use it to lift actions like `ReaderT.ask` to the appropriate position in the stack. And since @patrickt observed that for any monad `m`, `Algebra (Lift m) m` is definable with precisely the definition given for both `IO` and `Identity`, you really only need the one `Algebra` instance, too.

(In fact, you’d likely want to operate on monad transformers instead, since your signatures would otherwise grow quadratically:

```haskell
Lift (ExceptT e (StateT s (ReaderT r IO))) :+: Lift (StateT s (ReaderT r IO)) :+: Lift (ReaderT r IO) :+: Lift IO
```

which would be pretty obnoxious.)

That seems pretty central to what @lexi-lambda is exploring in [`eff`](https://github.com/hasura/eff/), and while I’m not ready to throw out our encoding of effects as data, it is _incredibly_ cool that _Effect Handlers in Scope_ also turns out to accommodate (essentially) `MonadBaseControl`—tho in hindsight this should be no surprise given Tom Schrijvers’ work on unlifting in _Monads, Zippers and Views: Virtualizing the Monad Stack_ and _The Monad Zipper_, as well as the fact that by its very nature, `Effect` is buying what `MonadTransControl` is selling.

Kind of a wonderful confluence of research, here, and I’m so delighted to have this!